common/src/main/java/io/pravega/common/io/serialization/VersionedSerializer.java

```diff
@@ -0,0 +1,407 @@
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void serialize(RevisionDataOutput dataOutput, TargetType o) throws IOException {
+        serialize(dataOutput.getBaseStream(), o);
+    }
+
+    /**
+     * Serializes the given object to the given OutputStream.
+     *
+     * @param stream The OutputStream to serialize to.
+     * @param o      The object to serialize.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void serialize(OutputStream stream, TargetType o) throws IOException {
+        // Wrap the given stream in a DataOutputStream, but make sure we don't close it, since we don't own it.
+        DataOutputStream dataOutput = stream instanceof DataOutputStream ? (DataOutputStream) stream : new DataOutputStream(stream);
+        val writeVersion = getFormat(writeVersion());
+
+        // Write Serialization Header.
+        dataOutput.writeByte(FORMAT_VERSION);
+        dataOutput.writeByte(writeVersion.getVersion());
+        dataOutput.writeByte(writeVersion.getRevisions().size());
+
+        // Write each Revision for this Version, in turn.
+        for (val r : writeVersion.getRevisions()) {
+            dataOutput.writeByte(r.getRevision());
+            try (val revisionOutput = RevisionDataOutputStream.wrap(stream)) {
+                r.getWriter().accept(o, revisionOutput);
+            }
+        }
+    }
+
+    /**
+     * Verifies that the given condition is true. If not, throws a SerializationException.
+     */
+    private void ensureCondition(boolean condition, String messageFormat, Object... args) throws SerializationException {
+        if (!condition) {
+            throw new SerializationException(String.format(messageFormat, args));
+        }
+    }
+
+    /**
+     * Deserializes data from the given RevisionDataInput into the given object. This overload is usually invoked for
+     * deserializing nested classes or collections.
+     *
+     * @param dataInput The RevisionDataInput to deserialize from.
+     * @param target    The target object to apply the deserialization to.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void deserialize(RevisionDataInput dataInput, ReaderType target) throws IOException {
+        deserialize(dataInput.getBaseStream(), target);
+    }
+
+    /**
+     * Deserializes data from the given InputStream into the given object.
+     *
+     * @param stream The InputStream to deserialize from.
+     * @param target The target object to apply the deserialization to.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void deserialize(InputStream stream, ReaderType target) throws IOException {
+        DataInputStream dataInput = stream instanceof DataInputStream ? (DataInputStream) stream : new DataInputStream(stream);
+        byte formatVersion = dataInput.readByte();
+        ensureCondition(formatVersion == FORMAT_VERSION, "Unsupported format version %d.", formatVersion);
```

Comment:
for backward compatibility, should we check for `<=` so that we attempt to deserialize latest and older versions?